### ISSUE TITLE
Implement API for getting conversion script information

### ISSUE THREAD

**Issue by perost** [OpenModelica/OpenModelica#8111]
OMEdit needs information about the available conversions in order to use `convertPackage` to convert user models. One suggestion is to implement something like `getConversionScripts(Modelica, "3.2.3")` (or maybe `getAvailableConversions` is a better name?) that returns a list of compatible versions of that library that the user has installed and the conversion scripts associated with them.

One possible complication is that we want to avoid parsing the whole library if we can, we only need the conversion annotation so in most cases it's probably enough to just parse the `package.mo` file.

**Comment by adeas31** [comment]
One basic use case that we discussed is lets say the user wants to convert the following model,

```
model M
  Modelica.SIunits.Angle a;
equation

annotation(
    uses(Modelica(version = "3.2.3")));
end M;
```

In that case OMEdit can read the uses annotation of the user model and call the new api to get the versions and path to conversion scripts. Something like,  `getConversionScripts(Modelica, "3.2.3")` (or whatever the name of the new api is) and get back `{"4.0.0", "path to conversion script"}`

**Comment by sjoelund** [comment]
There is an openmodelica metadata JSON file for libraries installed by the package manager which contains the information you need.

**Comment by sjoelund** [comment]
If you need the functionality, I think extending `getAvailablePackageVersions` to return a list of packages that have conversion scripts as a second output.

**Comment by perost** [comment]
> There is an openmodelica metadata JSON file for libraries installed by the package manager which contains the information you need.

It doesn't contain the path to the conversion script though, right?

**Comment by sjoelund** [comment]
> It doesn't contain the path to the conversion script though, right?

It could perhaps be added, but would require changing the format of the package index or adding redundant information.
I think loading the library itself before starting conversion script isn't too bad of a requirement. Once it's loaded, the regular API to get string annotations probably work.

**Comment by perost** [comment]
> > It doesn't contain the path to the conversion script though, right?
> 
> It could perhaps be added, but would require changing the format of the package index or adding redundant information. I think loading the library itself before starting conversion script isn't too bad of a requirement. Once it's loaded, the regular API to get string annotations probably work.

I guess we can use the package manager to get a list of only the libraries that are compatible with conversion. But then we need to get the path to the conversion script for each of those, since we want to allow the user to choose which version to upgrade to in case there's multiple.

But for libraries split into multiple files it should be sufficient to just parse the `package.mo` file in order to get the conversion annotation, and packages not split into multiple files are probably relatively small anyway.

**Comment by sjoelund** [comment]
> I guess we can use the package manager to get a list of only the libraries that are compatible with conversion. But then we need to get the path to the conversion script for each of those, since we want to allow the user to choose which version to upgrade to in case there's multiple.

You don't need to get the path to each of those. You only need to make the user choose version from the list and then convert based on this (load package.mo, parse annotations, find the correct script).

**Comment by perost** [comment]
> > I guess we can use the package manager to get a list of only the libraries that are compatible with conversion. But then we need to get the path to the conversion script for each of those, since we want to allow the user to choose which version to upgrade to in case there's multiple.
> 
> You don't need to get the path to each of those. You only need to make the user choose version from the list and then convert based on this (load package.mo, parse annotations, find the correct script).

That's true, I was stuck on our original plan. So we make `getAvailablePackageVersions` also return a list of packages that are compatible with conversion, then we just need some API for fetching the conversion script for a single library version.

> > and packages not split into multiple files are probably relatively small anyway.
> 
> They're really not in some cases :)

I like to at least pretend :) There's not much we can do in that case anyway, we need to parse at least the main file of the library regardless of how large it is.

**Comment by adeas31** [comment]
Meanwhile I was testing the conversion API and it does seem to work but does not update the uses annotation. For example,

```
loadString("model M
  Modelica.SIunits.Angle a;
equation
annotation(
    uses(Modelica(version = \"3.2.3\")));
end M;");
getErrorString();
list(M);
getErrorString();
convertPackage(M, "C:/Users/adeas31/AppData/Roaming/.openmodelica/libraries/Modelica 4.0.0+maint.om/Resources/Scripts/Conversion/ConvertModelica_from_3.2.3_to_4.0.0.mos");
getErrorString();
list(M);
getErrorString();
```

Run the above script. Update the hard coded conversion script path. The output of the script is,

```
true
""
"model M
  Modelica.SIunits.Angle a;
equation

  annotation(
    uses(Modelica(version = \"3.2.3\")));
end M;"
""
true
""
"model M
  Modelica.Units.SI.Angle a;
equation

  annotation(
    uses(Modelica(version = \"3.2.3\")));
end M;"
""
```
Note that the `Modelica.SIunits.Angle` is updated to `Modelica.Units.SI.Angle` but the uses annotation is same as `3.2.3`. This means the next time I load the class it will load MSL 3.2.3 and the class will not work with that version.

**Comment by perost** [comment]
> Meanwhile I was testing the conversion API and it does seem to work but does not update the uses annotation.

Good point. The conversion script itself does not contain any information on what version it converts to, so we might need to pass that as an argument to `convertPackage` since OMEdit should now that. We could try and deduce it from the filename, but that doesn't seem like a very good solution.

**Comment by sjoelund** [comment]
Yeah, for my script I just knew we always do MSL 3 to 4.0.0: https://github.com/OpenModelica/OpenModelicaLibraryTesting/blob/master/conversionscript.py#L100

Might be easier to force parsing the library to get the version annotation

**Comment by perost** [comment]
I've now opened #8169 that should solve this issue, but it's not quite ready to be merged yet because I need to update the library testing script a bit and also the test cases we have.

Due to the issue of having to update the uses-annotation that @adeas31 pointed out I changed the API a bit. I added a new function:
```Modelica
function convertPackageToLibrary
  input TypeName packageToConvert;
  input TypeName library;
  input String libraryVersion;
  output Boolean success;
end convertPackageToLibrary;
```
This function loads the requested library and looks up the conversion script (or scripts in case it needs multiple steps to convert) itself, converts the package, and then updates the uses-annotation. The old `convertPackage` function is still available, but I renamed it `runConversionScript` to better reflect what it does.

I also implemented `getAvailablePackageConversionsFrom` that OMEdit can use to list which versions of a library that provide conversions from a given version. I.e. `getAvailablePackageConversionsFrom(Modelica, "3.2.3")` returns `{"4.0.0"}`. I don't know if OMEdit knows which version of a specific library a model is using, but I can implement an API function for that if needed. Also, this returns all versions that the package manager knows about regardless of if they're installed or not, so OMEdit should probably filter it against the list of installed packages or maybe offer to install the selected package if it isn't already installed.

And in a brief state of confusion I first implemented `getAvailablePackageConversionsTo`, which tells you which versions of a library that can be converted to the request version. I'm not sure how useful that actually is, but I included it anyway since I'd already implemented it when I realised it wasn't what I needed.

**Comment by sjoelund** [comment]
`getVersion(Modelica)` should work already :)

**Comment by perost** [comment]
> `getVersion(Modelica)` should work already :)

Only if we assume that the loaded version of Modelica is also the one the model that we want to convert is using.

**Comment by sjoelund** [comment]
Oh, so you meant checking for uses annotations? `getUses(Bla)`

**Comment by perost** [comment]
> Oh, so you meant checking for uses annotations? `getUses(Bla)`

Yeah, that works.

**Comment by perost** [comment]
Also, `convertPackageToLibrary` will load the library to convert to, but I'm not sure what to do with it after the conversion. Right now I keep it loaded, overwriting any other version of the library that was loaded earlier. But we can easily discard it instead and not touch the loaded libraries if that's preferable.

There might also be some issues with loading dependencies, `CevalScript.loadModel` can silently load a new version of a library but will complain about libraries loaded via uses-annotations. I guess we don't actually need to load dependencies for the conversion, but if we want to keep the new library loaded we should also make sure the dependencies are correct.

**Comment by perost** [comment]
#8169 is now merged, and I also updated the conversion testing script to reflect the changes I made to the API.

**Comment by casella** [comment]
Should be done, feedback @adeas31 is appreciated

**Comment by adeas31** [comment]
See #8226

**Comment by casella** [comment]
> See #8226

@adease31, I'm confused. #8226 is about installing libraries, not converting them.

**Comment by adeas31** [comment]
No #8226 is indeed for running conversion script but I also fixed #8187 with it.

**Comment by perost** [comment]
The API is done and being used by OMEdit, so I'll close this and if there are any issues found we can open more specific issues for those.
